File: griffon-app/controllers/ru/dkolmogortsev/ControlPanelController.java
package ru.dkolmogortsev;

import griffon.core.artifact.GriffonController;
import griffon.inject.MVCMember;
import griffon.metadata.ArtifactProviderFor;
import griffon.transform.Threading;
import griffon.transform.Threading.Policy;
import java.time.Duration;
import java.util.List;
import java.util.stream.Collectors;
import javax.inject.Inject;
import org.codehaus.griffon.runtime.core.artifact.AbstractGriffonController;
import org.reactfx.util.FxTimer;
import org.reactfx.util.Timer;
import ru.dkolmogortsev.messages.StartTask;
import ru.dkolmogortsev.messages.TaskStopped;
import ru.dkolmogortsev.task.Task;
import ru.dkolmogortsev.task.TimeEntry;
import ru.dkolmogortsev.task.search.SearchFields;
import ru.dkolmogortsev.task.search.TaskSearcher;
import ru.dkolmogortsev.task.storage.TaskStorage;
import ru.dkolmogortsev.task.storage.TimeEntriesStorage;

@ArtifactProviderFor(GriffonController.class)
public class ControlPanelController extends AbstractGriffonController
{
    @MVCMember
    private ControlPanelModel model;

    @Inject
    private TaskSearcher searcher;

    @Inject
    private TaskStorage storage;

    @Inject
    private TimeEntriesStorage entriesStorage;

    private Timer backUpTimer;

    @Threading(Threading.Policy.INSIDE_UITHREAD_SYNC)
    public void start()
    {
        if (!model.isTaskStarted())
        {
            String description = model.taskDescriptionProperty().get();
            String nameUI = model.taskNameProperty().get();
            Task saved = storage.save(new Task(description, Task.NO_PROJECT));
            TimeEntry te = new TimeEntry(System.currentTimeMillis(), saved);
            entriesStorage.save(te);
            model.currentTimeEntryIdProperty().set(te.getId());
            initTimeEntryBackup(te);
            model.startTimer();
        }
        else{
            TimeEntry te = entriesStorage.get(model.currentTimeEntryIdProperty().get());
            te.stop();
            getApplication().getEventRouter().publishEvent(new TaskStopped(te.getId()));
            model.stopTimer();
        }
    }

    @Threading(Policy.OUTSIDE_UITHREAD)
    private void initTimeEntryBackup(TimeEntry te)
    {
        backUpTimer = FxTimer.runPeriodically(Duration.ofSeconds(10), () ->
        {
            te.updateDuration(model.getElapsedProperty());
            entriesStorage.save(te);
        });
    }

    @Threading(Threading.Policy.INSIDE_UITHREAD_ASYNC)
    public void search(String searchStr, SearchFields field)
    {
        List<String> search = searcher.search(searchStr, field).stream().map(it -> field.getMapper().apply(it))
                .collect(Collectors.toList());
        model.setList(search);
    }

    public void onStartTask(StartTask task){
        Task taskFromStorage = storage.getTask(task.getTaskId());
        model.taskDescriptionProperty().set(taskFromStorage.getDescription());
        model.taskNameProperty().set(taskFromStorage.getTaskName());
        start();
    }

}

File: griffon-app/controllers/ru/dkolmogortsev/ControlPanelController.kt
package ru.dkolmogortsev

import griffon.core.artifact.GriffonController
import griffon.inject.MVCMember
import griffon.metadata.ArtifactProviderFor
import griffon.transform.Threading
import griffon.transform.Threading.Policy
import org.codehaus.griffon.runtime.core.artifact.AbstractGriffonController
import org.reactfx.util.FxTimer
import org.reactfx.util.Timer
import ru.dkolmogortsev.events.TaskSelectedEvent
import ru.dkolmogortsev.events.TaskStartedEvent
import ru.dkolmogortsev.events.TaskStoppedEvent
import ru.dkolmogortsev.task.Task
import ru.dkolmogortsev.task.TimeEntry
import ru.dkolmogortsev.task.search.SearchFields
import ru.dkolmogortsev.task.search.TaskSearcher
import ru.dkolmogortsev.task.storage.TaskStorage
import ru.dkolmogortsev.task.storage.TimeEntriesStorage
import java.time.Duration
import javax.inject.Inject

@ArtifactProviderFor(GriffonController::class)
class ControlPanelController : AbstractGriffonController()
{
    @MVCMember
    lateinit var model: ControlPanelModel
    @Inject
    private lateinit var searcher: TaskSearcher
    @Inject
    private lateinit var storage: TaskStorage
    @Inject
    private lateinit var entriesStorage: TimeEntriesStorage
    private lateinit var backUpTimer: Timer
    override fun mvcGroupInit(args: MutableMap<String, Any>)
    {
        getApplication().eventRouter
                .addEventListener(TaskStartedEvent::class.java, { onStartTask(it!![0] as TaskStartedEvent) })
        getApplication().eventRouter
                .addEventListener(TaskSelectedEvent::class.java, { onTaskSelected(it!!.first() as TaskSelectedEvent) })
        super.mvcGroupInit(args)
    }

    @Threading(Threading.Policy.INSIDE_UITHREAD_SYNC)
    fun start()
    {
        if (!model.isTaskStarted)
        {
            val description = model.taskDescriptionProperty().get()
            val nameUI = model.taskNameProperty().get()
            val task = Task(description, nameUI)
            storage.save(task)
            val te = TimeEntry(System.currentTimeMillis(), task)
            entriesStorage.save(te)
            model.timeEntryId = te.id
            initTimeEntryBackup(te)
            model.startTimer()
        }
        else
        {
            stopBackgroundBackup()
            val te = entriesStorage.get(model.timeEntryId)
            te.updateDuration(model.getElapsedProperty())
            te.stop()
            entriesStorage.save(te)
            getApplication().eventRouter.publishEvent(TaskStoppedEvent(te.id))
            model.stopTimer()
        }
    }

    @Threading(Policy.OUTSIDE_UITHREAD)
    private fun initTimeEntryBackup(te: TimeEntry)
    {
        backUpTimer = FxTimer.runPeriodically(Duration.ofSeconds(10)) {
            te.updateDuration(model.getElapsedProperty())
            entriesStorage.save(te)
        }
    }

    @Threading(Policy.OUTSIDE_UITHREAD)
    private fun stopBackgroundBackup()
    {
        backUpTimer.stop()
    }

    @Threading(Threading.Policy.INSIDE_UITHREAD_ASYNC)
    fun search(searchStr: String, field: SearchFields)
    {
        model.setSearchResults(searcher.search(searchStr, field))
    }

    fun onStartTask(taskStartedEvent: TaskStartedEvent)
    {
        val taskFromStorage = storage.getTask(taskStartedEvent.taskId)
        model.taskDescriptionProperty().set(taskFromStorage.description)
        model.taskNameProperty().set(taskFromStorage.taskName)
        start()
    }

    fun onTaskSelected(taskSelectedEvent: TaskSelectedEvent)
    {
        val task = storage[taskSelectedEvent.taskId]
        model.taskDescriptionProperty().value = task.description
        model.taskNameProperty().value = task.taskName
        model.taskSelectedProperty.value = true
        model.taskSelectedProperty.value = false
    }
}